### ISSUE TITLE
Goals of SegmentStrategyTest unclear

### ISSUE THREAD

**Issue by slarse** [SpoonLabs/sorald#154]
The two tests in `SegmentStrategyTest.java` have pretty unclear goals. I changed the directory structure in the test resources, which caused a code coverage diff in `FirstFitSegmentDiffAlgorithm` ([see coverage diff here](https://codecov.io/gh/SpoonLabs/sorald/pull/144/changes)). The tests still pass, though, so I'm not certain what they're actually supposed to test, and there's no documentation to clarify it.

I'll try to figure it out, but if anyone has any suggestions please feel free to add them!

**Comment by slarse** [comment]
The package statement is also incorrect, doesn't match the directory structure.

**Comment by fermadeiral** [comment]
@henry-lp is the creator of the segment repair strategy.

Henry, could you please provide some information about it here?

**Comment by slarse** [comment]
@fermadeiral Thanks!

So, I've figured out the "fail" test, it just tests that the 0 is an invalid value for max files per segment:

https://github.com/SpoonLabs/sorald/blob/ea7d6264d5ad02eb52189fed7cda65ebac516a6d/src/test/java/sorald/SegmentStrategyTest.java#L39

I've also figured out the different execution path for the "success" test. It operates directly on the resource directory, with a max segment size of 3. https://github.com/SpoonLabs/sorald/blob/ea7d6264d5ad02eb52189fed7cda65ebac516a6d/src/test/java/sorald/SegmentStrategyTest.java#L19-L25

Now, with the current state of #144, the directories in the resources dir have significantly fewer files than before, and so this condition is never true: https://github.com/SpoonLabs/sorald/blob/ea7d6264d5ad02eb52189fed7cda65ebac516a6d/src/main/java/sorald/segment/FirstFitSegmentationAlgorithm.java#L26

Hence the coverage diff.

@henry-lp Could you explain what the purpose of the success test is?

Also, why is there a differentiation between a "file node" and a "directory node"? To me, it looks like a separation of the contents of a directory (the file node) and its name (the directory node). Why is this necessary?

Finally, could you in very general terms describe the segmentation strategy?

Your input would be hugely appreciated, it'd save me a lot of time :)

**Comment by henry-lp** [comment]
Hi @slarse  :),
> @henry-lp Could you explain what the purpose of the success test is?
The purpose is to verify that with segment size of 3 (< the number of files in `Constants.PATH_TO_RESOURCES_FOLDER`) Segment mode can find and repair a violation of rule 2116 (.i.e ArrayHashCodeAndToStringCheck) which is verified at the end by Sonar-Java that no more violations of such rule can be found. 

> Finally, could you in very general terms describe the segmentation strategy? It looks to me like it's a matter of putting all Java files into a list of sublists of size <= MAX_SEGMENT_SIZE, where each sublist is processed in one go. Correct?
Yes, the concept of the segment mode is to separate the input project into smaller pieces to analyze and repair for warnings. It is simple but can be proven to work well for Eclipse projects in terms of fixes and time performance in comparison to the default mode, which fails to produce any fixes at all once an error occur during a repair attempt.

> Also, why is there a differentiation between a "file node" and a "directory node"? To me, it looks like a separation of the contents of a directory (the file node) and its name (the directory node). Why is this necessary?
Given a directory you might have subfolders and some files. Each subfolder is a directory node, while all the files are putted in a single file node. 
The vision I had in mind when I implemented this segment mode is to mainly have a good data structure to accommodate possible different segmentation algorithms in the future other than the first fit segmentation. In this case, I think a tree is a good and natural data structure to represent the folders and files of the input project. Also, given a directory I need the information on exactly how many files are in total - files of the current folder + all files of all its subfolder. My initial idea was to only fit as many folders as we can given a max segment size. I later found out that the experiment result would be just as well if you are just putting as many files of the same folder into the same segment and go to the next segment once the previous segment is full. 
Regardless it indicates that we might not need very complicated approach to handle bigger projects when repairing with Sorald.

Long story short, if you think the tree is redundant feel free to swap it out with something else to your liking. Hopefully, this wall of text is helpful for your work :) . Feel free to pose a follow-up question if there are any.

**Comment by slarse** [comment]
Hi @henry-lp , thanks for the elaborate answer!

> The purpose is to verify that with segment size of 3 (< the number of files in Constants.PATH_TO_RESOURCES_FOLDER) Segment mode can find and repair a violation of rule 2116 (.i.e ArrayHashCodeAndToStringCheck) which is verified at the end by Sonar-Java that no more violations of such rule can be found.

Right, got it. Then I just need to harden the test case a bit (i.e. by checking the precondition) and make a little note of that. The problem now was that I made the precondition of `3 < num files in resources dir` false as part of a refactoring, but the test still passed without a problem.

> Yes, the concept of the segment mode is to separate the input project into smaller pieces to analyze and repair for warnings.

Great, then that's clear to me.

> if you think the tree is redundant feel free to swap it out with something else to your liking

No, I think a tree is the proper and intuitive abstraction here. It's modelling a file system, after all.

> Each subfolder is a directory node, while all the files are putted in a single file node.

Right, so this is the one thing that had me a bit puzzled about the tree structure. Essentially, as each node has one list of subdirectories (child nodes) and one list of files in separate data structures, there's no need to distinguish between "directory" nodes and "file" nodes. They could all just be directory nodes, removing the indirection between the directory and its contents. My question was essentially, is there any particular reason for this indirection, or is it something that just kind of happened?

**Comment by henry-lp** [comment]
Hi @slarse :), 

>.. .My question was essentially, is there any particular reason for this indirection, or is it something that just kind of happened?

I think this confusion occurs due to me being too lazy to separate the implementation of a file node and a directory node, since everything is just implemented as a node currently. According to my definitions, a directory node should not have any files, while a file node should not have any child (subdirectory). Then, the file node is allowed to be separated to smaller file nodes with each node contains a subset of the original file set so that the segment under construction can be filled completely. Basicly, this is just my way of representing the file structure I want to work with when I started to hypothesize about possible different algorithms back then. 

In conclusion, I agree that for now you can also have a less distinct definition by seeing them all as directory nodes. By doing that, a directory node now either contains only files and no subdirectories or no files and only subdirectories or both files and subdirectories. If you go in that direction then it is in principle only a matter of removing the comments in the `Node.java` end rename it to `DirNode.java` or regards it as a directory node only from now only.

**Comment by slarse** [comment]
@henry-lp Sorry, I just kind of forgot to answer here!

> Then, the file node is allowed to be separated to smaller file nodes with each node contains a subset of the original file set so that the segment under construction can be filled completely.

Ah, yes, when you put it like that I see why you did it like that. I understand the intent of the design now.

The thing I'm not super comfortable with in the current implementation is that there is no type distinction between a file node and a directory node, it's just convention, which in my experience can lead to mistakes and confusion. So I think the best way forward is to either remove make all nodes directory nodes, as you suggested, or actually make file nodes and directory nodes be different types of nodes. I.e. either remove the convention, or enforce it.

**Comment by henry-lp** [comment]
> @henry-lp Sorry, I just kind of forgot to answer here!
@slarse no worries :) . 
> make file nodes and directory nodes be different types of nodes. I.e. either remove the convention, or enforce it.
Then I'll start working on a PR for this refactoring. I think this would be best for me since it is related to the content of my thesis. If somehow in the future we think the other solution (i.e., calling all nodes as directory nodes) is better than we can refactor it to that.  Do you already have some plan in store for this refactoring work or shall I work on a PR for this? If it's the latter then I'll start today after my work about at 17:00.

**Comment by slarse** [comment]
@henry-lp This is by no means urgent, we have other priorities at the moment. As far as I can tell, the segmentation strategy is working fine, so what should precede any refactoring is a hardening of the tests to actually verify that they test what they should (which at the moment, they don't). As I now understand the intentions of the design and the tests, that's no problem. Furthermore, as the segmentation strategy is so very loosely coupled to the rest of the code base, postponing a refactoring imposes essentially no technical debt.

To be clear: there is nothing we're doing right now that requires or even benefits from a refactoring of the segmentation strategy. We just kind of slid into that discussion by accident. Possibly my fault :)

If I were you, I'd focus on finishing the thesis. Whether or not we change the details of the implementation of the segmentation strategy down the road really is of no concern to the validity of your thesis, as the general principle of the segmentation still applies.

**Comment by henry-lp** [comment]
@slarse point taken :) . This can be left for another day then.

**Comment by slarse** [comment]
@henry-lp Do you have any example projects off the top of your head that are too large to process with the default strategy?

**Comment by henry-lp** [comment]
@slarse it has been a while, so my memory is a bit hazy, but you can try these three:
* https://github.com/eclipse/eclipse.platform.ui
* https://github.com/eclipse/rdf4j
* https://github.com/eclipse/eclipse-collections

I think I got GC overhead limit exception on those.

**Comment by slarse** [comment]
Thanks!
